File: src/designer/designer_packages/elsa-workflows-designer/src/modules/switch/sequence/switch-port-provider.ts
import 'reflect-metadata';
import {Service} from "typedi";
import {Activity, Port, PortType} from "../../../models";
import {SwitchActivity, SwitchCase} from "./models";
import {PortProvider, PortProviderContext} from "../../../services";

@Service()
export class SwitchPortProvider implements PortProvider {

  getOutboundPorts(context: PortProviderContext): Array<Port> {
    const activity = context.activity as SwitchActivity;

    if (activity == null)
      return [];

    const cases = activity.cases ?? [];
    const ports: Port[] = cases.map(x => ({name: x.label, displayName: x.label, type: PortType.Embedded}));
    const defaultPort: Port = {name: 'default', displayName: 'Default', type: PortType.Embedded};
    return [...ports, defaultPort];
  }

  resolvePort(portName: string, context: PortProviderContext): Activity | Array<Activity> {
    const activity = context.activity as SwitchActivity;

    if (portName == 'default')
      return activity.default;

    const cases: Array<SwitchCase> = activity.cases ?? [];
    const caseItem = cases.find(x => x.label == portName);

    if (!caseItem)
      return null;

    return caseItem.activity;
  }

  assignPort(portName: string, activity: Activity, context: PortProviderContext) {
    const switchActivity = context.activity as SwitchActivity;

    if (portName == 'default') {
      switchActivity.default = activity;
      return;
    }

    const cases: Array<SwitchCase> = switchActivity.cases ?? [];
    const caseItem = cases.find(x => x.label == portName);

    if (!caseItem)
      return;

    caseItem.activity = activity;
  }
}

File: src/studio/modules/Elsa.Studio.ActivityPortProviders/Providers/SwitchPortProvider.cs
using System.Text.Json.Nodes;
using Elsa.Api.Client.Extensions;
using Elsa.Api.Client.Resources.ActivityDescriptors.Enums;
using Elsa.Api.Client.Resources.ActivityDescriptors.Models;
using Elsa.Studio.Workflows.Domain.Contexts;
using Elsa.Studio.Workflows.Domain.Providers;

namespace Elsa.Studio.ActivityPortProviders.Providers;

/// <summary>
/// Provides ports for the Switch activity based on its cases.
/// </summary>
public class SwitchPortProvider : ActivityPortProviderBase
{
    /// <inheritdoc />
    public override bool GetSupportsActivityType(PortProviderContext context) => context.ActivityDescriptor.TypeName == "Elsa.Switch";

    /// <inheritdoc />
    public override IEnumerable<Port> GetPorts(PortProviderContext context)
    {
        var cases = GetCases(context.Activity).ToList();

        foreach (var @case in cases)
        {
            var label = GetLabel(@case);
            yield return new Port
            {
                Name = label,
                DisplayName = label,
                Type = PortType.Embedded,
            };
        }
        
        if (cases.Any(x => GetLabel(x) == "Default"))
            yield break;
        
        yield return new Port
        {
            Name = "Default",
            DisplayName = "Default",
            Type = PortType.Embedded
        };
    }

    /// <inheritdoc />
    public override JsonObject? ResolvePort(string portName, PortProviderContext context)
    {
        if (portName == "Default")
        {
            return context.Activity.GetProperty<JsonObject>("default");
        }
        
        var cases = GetCases(context.Activity);
        var @case = cases.FirstOrDefault(x => GetLabel(x) == portName);

        return @case != null ? GetActivity(@case) : null;
    }

    /// <inheritdoc />
    public override void AssignPort(string portName, JsonObject activity, PortProviderContext context)
    {
        if(portName == "Default")
        {
            context.Activity.SetProperty(activity, "default");
            return;
        }
        
        var cases = GetCases(context.Activity).ToList();
        var @case = cases.FirstOrDefault(x => GetLabel(x) == portName);

        if (@case == null)
            return;

        SetActivity(@case, activity);
    }

    /// <inheritdoc />
    public override void ClearPort(string portName, PortProviderContext context)
    {
        if(portName == "Default")
        {
            context.Activity.SetProperty(null, "default");
            return;
        }
        
        var cases = GetCases(context.Activity).ToList();
        var @case = cases.FirstOrDefault(x => GetLabel(x) == portName);

        if (@case == null)
            return;

        SetActivity(@case, null);
    }

    private string GetLabel(JsonObject @case) => @case.GetProperty("label")?.GetValue<string>()!;
    private JsonObject? GetActivity(JsonObject @case) => @case.GetProperty("activity")?.AsObject();
    private void SetActivity(JsonObject @case, JsonObject? activity) => @case.SetProperty(activity, "activity");

    private static IEnumerable<JsonObject> GetCases(JsonObject switchActivity)
    {
        return switchActivity.GetProperty("cases")?.AsArray().AsEnumerable().Cast<JsonObject>() ?? new List<JsonObject>();
    }
}